k2/csrc/pinned_context.cu

```diff
@@ -0,0 +1,357 @@
+   */
+  void RecordEvent(cudaStream_t stream, void *ptr) {
+    NVTX_RANGE(K2_FUNC);
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto it = blocks_.find(ptr);
+
+    if (it == blocks_.end()) {
+      // this pointer is not returned by `Malloc`, ignore it.
+      return;
+    }
+
+    Block &block = it->second;
+    K2_CHECK(block.allocated)
+        << "RecordEvent is called with a block that has not been allocated!";
+    block.streams.insert(stream);
+  }
+
+ private:
+  cudaError_t InsertEvents(Block &block) {
+    NVTX_RANGE(K2_FUNC);
+    // InsertEvents is called from `Free`, which has already held the mutex.
+    std::unordered_set<cudaStream_t> streams(std::move(block.streams));
+    for (auto it = streams.begin(); it != streams.end(); ++it) {
+      cudaEvent_t event;
+      cudaError_t err =
+          cudaEventCreateWithFlags(&event, cudaEventDisableTiming);
+      if (err != cudaSuccess) return err;
+
+      err = cudaEventRecord(event, *it);
+      if (err != cudaSuccess) return err;
+
+      ++block.event_count;
+      cuda_events_.emplace_back(event, block.ptr);
+    }
+    return cudaSuccess;
+  }
+
+  /* Process events in `cuda_events_`.
+
+     If the events of a block have all been processed, this block
+     is put into `available_` and is ready for reuse.
+
+     If `cudaEventQuery()` returns `cudaErrorNotReady`, it
+     returns immediately.
+
+     @return `cudaSuccess` on success; on error, it returns a
+              cuda error code.
+   */
+  cudaError_t ProcessEvents() {
+    NVTX_RANGE(K2_FUNC);
+    // InsertEvents is called from `Malloc` and `Free`,
+    // which has already held the mutex.
+    while (!cuda_events_.empty()) {
+      auto &e = cuda_events_.front();
+      cudaEvent_t event = e.first;
+
+      cudaError_t err = cudaEventQuery(event);
+      if (err == cudaErrorNotReady) break;
+
+      if (err != cudaSuccess) return err;
+
+      err = cudaEventDestroy(event);
+      if (err != cudaSuccess) return err;
+
+      Block &block = blocks_.at(e.second);
+      --block.event_count;
+
+      if (block.event_count == 0 && !block.allocated) available_.insert(block);
+
+      cuda_events_.pop_front();
+    }
+    return cudaSuccess;
+  }
+
+ private:
+  // It contains all blocks allocated by Malloc.
+  std::unordered_map<void *, Block> blocks_;
+
+  using Compare = bool (*)(const BlockSize &, const BlockSize &);
+  // It contains all free blocks **sorted** by block size in increasing order.
+  std::set<BlockSize, Compare> available_;
+
+  // outstanding cuda events
+  std::deque<std::pair<cudaEvent_t, void *>> cuda_events_;
+
+  // to protect `blocks_`, `available_` and `cuda_events_` being accessed
+  // from multiple threads
+  std::mutex mutex_;
+};
+
+static PinnedAllocator *GetPinnedAllocator() {
+  static std::once_flag init_flag;
+  static PinnedAllocator *allocator = nullptr;
+
+  std::call_once(init_flag, []() {
+    // it is never freed.
+    allocator = new PinnedAllocator;
```

Comment:
also add documentation in the section of `valgrind` of INSTALL.md about this?

Reply:
I am wondering about the value of `valgrind` in k2. 

It shows lots of warnings from `libcuda.so` and we can do nothing about them.
`valgrind` only shows that there are some memory being leaked.
 
What should we document in INSTALL.md?